### PR TITLE
Bugfix UnitInput being used for SI units

### PR DESCRIPTION
**Pull request by mantis-toboggan-md**
#4417 - this PR fixes/simplifies/clarifies UnitInput's ability to parse and format SI values. There are some kubernetes resources with fields that accept either an integer base unit or a string with an exponential modifier; eg ResourceQuota memory reservation can be specified as an integer 524288 (Bytes) or string "0.5Mi" (Bytes) or "512Ki". UnitInput had some code that looked intended to deal with these type of values but it wasn't quite working, and was hard to follow...

 Namespace/workload resource limits used a different component to parse SI values, Projects used the vanilla UnitInput (simply doesn't work right in that case, hence the bug report), and the Harvester team added two more props to UnitInput to deal with it in a different way. All that now works in the same way with this PR.
 
This updated UnitInput accepts either a plain integer or a string with SI modifier, parses the string into a base unit integer, displays it with whatever exponential modifier is specified by `inputExponent` and outputs either a string with exponential modifier, or an integer the equivalent value converted to base units.
 
 UnitInput is also used in a ton of places as a simple integer input with a unit label displayed next to it. Those uses should be unchanged by this PR.

### CODE REVIEW

**Review by richard-cox — CHANGES_REQUESTED** [review]
This is a brave PR!

Haven't deep dived into every line of code, though hit a few snags when testing
- Create Namespace - Looks like this uses the project's container default resource limits instead of the namespace default limits. I'm not sure this is related to this PR, if not let me know and I'll create a PR
- Create HPA - fails after trying to submit a Metric of Memory-> Average Value & 80 (`value POSTed is `80MiMi`
- Create Harvester Volume - It looks like it submits the wrong value... but shows what's entered. For example entering 1024 (shows units of MiB) submits `"storage":"1024Gi"`. The volume list page shows `1Ti`, going back to the edit page shows 1024 MiB. When you're ready to look at this one let me know, I've got my publicly available Rancher with a Harvester attached
- Create Harvester VM - There was an `Cannot read properties of undefined (reading 'metadata')` error. Pretty sure this isn't related to PR but might have been fixed with a recent merge of HV updates. Could you rebase and I can test again?

These would be perfect for some unit tests. I've added myself as a reviewer to the unit test PR and will have a look this week.

File: components/ContainerResourceLimit.vue [hunk 1]
```diff
@@ -43,30 +41,9 @@ export default {
...
-      limitsMemory: parsed.limitsMemory ? formatSi(parsed.limitsMemory, {
-        minExponent: 2, maxExponent: 2, addSuffix: false, increment: 1024,
-      }) : null,
-      requestsCpu:  parsed.requestsCpu ? formatSi(parsed.requestsCpu, {
-        minExponent: 1, maxExponent: 1, addSuffix: false, increment: 1 / 1000,
-      }) : null,
-      requestsMemory: parsed.requestsMemory ? formatSi(parsed.requestsMemory, {
-        minExponent: 2, maxExponent: 2, addSuffix: false, increment: 1024,
-      }) : null,
-      viewMode: _VIEW,
```
[the remark this answers is not in the document]
Comment (mantis-toboggan-md) [reply components/ContainerResourceLimit.vue:66]: re-added

**Comment by mantis-toboggan-md** [comment]
> * Create Namespace - Looks like this uses the project's container default resource limits instead of the namespace default limits. I'm not sure this is related to this PR, if not let me know and I'll create a PR

If there is a bug there it wasn't introduced in this PR, but it also appears to be working the same as it does in Ember. I just tried making namespaces in each and they end up with the same defaults, though Ember allows you to specify resourcequotas different from the project default:
Here's via vue:
```
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    cattle.io/status: '{"Conditions":[{"Type":"ResourceQuotaValidated","Status":"True","Message":"","LastUpdateTime":"2021-11-17T15:55:00Z"},{"Type":"ResourceQuotaInit","Status":"True","Message":"","LastUpdateTime":"2021-11-17T15:55:00Z"},{"Type":"InitialRolesPopulated","Status":"True","Message":"","LastUpdateTime":"2021-11-17T15:55:00Z"}]}'
    field.cattle.io/containerDefaultResourceLimit: '{"limitsCpu":"12m","limitsMemory":"23Mi","requestsCpu":"34m","requestsMemory":"45Mi"}'
    field.cattle.io/projectId: local:p-h989z
    field.cattle.io/resourceQuota: '{"limit":{"requestsCpu":"30m","requestsMemory":"50Mi","limitsCpu":"20m","limitsMemory":"40Mi"}}'
    lifecycle.cattle.io/create.namespace-auth: "true"
  creationTimestamp: "2021-11-17T15:54:59Z"
  finalizers:
  - controller.cattle.io/namespace-auth
  labels:
    field.cattle.io/projectId: p-h989z
...
```
And Ember:
```
apiVersion: v1
kind: Namespace
metadata:
  annotations:
    cattle.io/status: '{"Conditions":[{"Type":"InitialRolesPopulated","Status":"True","Message":"","LastUpdateTime":"2021-11-17T15:54:49Z"},{"Type":"ResourceQuotaValidated","Status":"True","Message":"","LastUpdateTime":"2021-11-17T15:54:48Z"},{"Type":"ResourceQuotaInit","Status":"True","Message":"","LastUpdateTime":"2021-11-17T15:54:48Z"}]}'
    field.cattle.io/containerDefaultResourceLimit: '{"limitsCpu":"12m","limitsMemory":"23Mi","requestsCpu":"34m","requestsMemory":"45Mi"}'
    field.cattle.io/creatorId: user-c5r4d
    field.cattle.io/projectId: local:p-h989z
    field.cattle.io/resourceQuota: '{"limit":{"limitsCpu":"20m","limitsMemory":"40Mi","requestsCpu":"30m","requestsMemory":"50Mi"}}'
    lifecycle.cattle.io/create.namespace-auth: "true"
  creationTimestamp: "2021-11-17T15:54:47Z"
  finalizers:
  - controller.cattle.io/namespace-auth
  labels:
    cattle.io/creator: norman
    field.cattle.io/projectId: p-h989z
...
```

> * Create HPA - fails after trying to submit a Metric of Memory-> Average Value & 80 (value POSTed is `80MiMi`)

Fixed

> * Create Harvester Volume - It looks like it submits the wrong value... but shows what's entered. For example entering 1024 (shows units of MiB) submits `"storage":"1024Gi"`. The volume list page shows `1Ti`, going back to the edit page shows 1024 MiB. When you're ready to look at this one let me know, I've got my publicly available Rancher with a Harvester attached

I didn't see this bug exactly,  but it was indeed broken. Should be fixed now.

> * Create Harvester VM - There was an `Cannot read properties of undefined (reading 'metadata')` error. Pretty sure this isn't related to PR but might have been fixed with a recent merge of HV updates. Could you rebase and I can test again?

This appears to be fixed...I did notice when I make a harvester VM then go back to edit it the memory value in the first tab has changed. I double-checked and that behavior is the same in master...not really sure if it's intentional I don't think I know enough about harvester. Looking at my network tab, here's what the UI sent:
```
domain: {machine: {type: ""}, cpu: {cores: 2, sockets: 1, threads: 1},…}
cpu: {cores: 2, sockets: 1, threads: 1}
devices: {inputs: [{bus: "usb", name: "tablet", type: "tablet"}],…}
machine: {type: ""}
memory: {guest: "9.9Gi"}
resources: {requests: {memory: "10Gi", cpu: 2}}
requests: {memory: "10Gi", cpu: 2}
```

And the response:
```
domain: {cpu: {cores: 2, sockets: 1, threads: 1}, devices: {,…}, machine: {type: "q35"},…}
cpu: {cores: 2, sockets: 1, threads: 1}
devices: {,…}
machine: {type: "q35"}
memory: {guest: "10630044057600m"}
guest: "10630044057600m"
resources: {limits: {cpu: "2", memory: "10Gi"}, requests: {cpu: "125m", memory: "7158278826"}} ---> this is about 6.6GiB
limits: {cpu: "2", memory: "10Gi"}
requests: {cpu: "125m", memory: "7158278826"}
```

So I specified a resource request limit of 10Gi and a VM with a request limit of 7158278826 aka about 6.6GiB was returned. Seems odd but idk

> These would be perfect for some unit tests. I've added myself as a reviewer to the unit test PR and will have a look this week.
...I'll have to update those to account for changes here

**Review by richard-cox — APPROVED** [review]
I'll jump in a merge this one so that the Harvester team can fix their issue given the latest code
